File: VB/HeadFirstPatterns/DoFactory.HeadFirst/DoFactory.HeadFirst.Compound.Decorator/DuckSimulator.vb
Imports Microsoft.VisualBasic
Imports System

Namespace DoFactory.HeadFirst.Combining.Decorator
	Friend Class DuckSimulator
        Shared Sub Main(ByVal args() As String)

            Dim simulator As New DuckSimulator()
            simulator.Simulate()

        End Sub

        Private Sub Simulate()

            Dim mallardDuck As IQuackable = New QuackCounter(New MallardDuck())
            Dim redheadDuck As IQuackable = New QuackCounter(New RedheadDuck())
            Dim duckCall As IQuackable = New QuackCounter(New DuckCall())
            Dim rubberDuck As IQuackable = New QuackCounter(New RubberDuck())
            Dim gooseDuck As IQuackable = New GooseAdapter(New Goose())

            Console.WriteLine("Duck Simulator: With Decorator")

            Simulate(mallardDuck)
            Simulate(redheadDuck)
            Simulate(duckCall)
            Simulate(rubberDuck)
            Simulate(gooseDuck)

            Console.WriteLine("The ducks quacked " & QuackCounter.Quacks & " times")

            ' Wait for user
            Console.ReadKey()
        End Sub

		Private Sub Simulate(ByVal duck As IQuackable)
			duck.Quack()
		End Sub
	End Class

	#Region "Ducks "

	Public Interface IQuackable
		Sub Quack()
	End Interface

	Public Class RubberDuck
		Implements IQuackable
		Public Sub Quack() Implements IQuackable.Quack
			Console.WriteLine("Squeak")
		End Sub

		Public Overrides Function ToString() As String
			Return "Rubber Duck"
		End Function
	End Class

	Public Class RedheadDuck
		Implements IQuackable
		Public Sub Quack() Implements IQuackable.Quack
			Console.WriteLine("Quack")
		End Sub
	End Class

	Public Class MallardDuck
		Implements IQuackable
		 Public Sub Quack() Implements IQuackable.Quack
			Console.WriteLine("Quack")
		 End Sub

		Public Overrides Function ToString() As String
			Return "Mallard Duck"
		End Function
	End Class

	Public Class DuckCall
		Implements IQuackable
		 Public Sub Quack() Implements IQuackable.Quack
			Console.WriteLine("Kwak")
		 End Sub

		Public Overrides Function ToString() As String
			Return "Duck Call"
		End Function
	End Class

	Public Class DecoyDuck
		Implements IQuackable
		 Public Sub Quack() Implements IQuackable.Quack
			Console.WriteLine("<< Silence >>")
		 End Sub

		Public Overrides Function ToString() As String
			Return "Decoy Duck"
		End Function
	End Class

	#End Region

	#Region "QuackCounter"

	Public Class QuackCounter
		Implements IQuackable
		Private _duck As IQuackable

		Public Sub New(ByVal duck As IQuackable)
			Me._duck = duck
		End Sub

		Public Sub Quack() Implements IQuackable.Quack
			_duck.Quack()
			Quacks += 1
		End Sub

        Private Shared _quacks As Integer
		Public Shared Property Quacks() As Integer
			Get
                Return _quacks
			End Get
			Private Set(ByVal value As Integer)
                _quacks = value
			End Set
		End Property

		Public Overrides Function ToString() As String
			Return CType(_duck, Object).ToString()
		End Function
	End Class

	#End Region

	#Region "Goose "

	Public Class GooseAdapter
		Implements IQuackable
		Private _goose As Goose

		Public Sub New(ByVal goose As Goose)
			Me._goose = goose
		End Sub

		Public Sub Quack() Implements IQuackable.Quack
			_goose.Honk()
		End Sub

		Public Overrides Function ToString() As String
			Return "Goose pretending to be a Duck"
		End Function
	End Class

	Public Class Goose
		Public Sub Honk()
			Console.WriteLine("Honk")
		End Sub

		Public Overrides Function ToString() As String
			Return "Goose"
		End Function
	End Class
	#End Region
End Namespace

File: CSharp/HeadFirstPatterns/DoFactory.HeadFirst/DoFactory.HeadFirst.Compound.Decorator/DuckSimulator.cs
using System;

namespace DoFactory.HeadFirst.Combining.Decorator
{
    class DuckSimulator
    {
        static void Main(string[] args)
        {
            DuckSimulator simulator = new DuckSimulator();
            simulator.Simulate();
        }

        void Simulate() 
        {
            IQuackable mallardDuck = new QuackCounter(new MallardDuck());
            IQuackable redheadDuck = new QuackCounter(new RedheadDuck());
            IQuackable duckCall    = new QuackCounter(new DuckCall());
            IQuackable rubberDuck  = new QuackCounter(new RubberDuck());
            IQuackable gooseDuck   = new GooseAdapter(new Goose());

            Console.WriteLine("Duck Simulator: With Decorator");

            Simulate(mallardDuck);
            Simulate(redheadDuck);
            Simulate(duckCall);
            Simulate(rubberDuck);
            Simulate(gooseDuck);

            Console.WriteLine("The ducks quacked " + 
                QuackCounter.Quacks + " times");

            // Wait for user
            Console.ReadKey();
        }

        void Simulate(IQuackable duck) 
        {
            duck.Quack();
        }    
    }

    #region Ducks 

    public interface IQuackable 
    {
        void Quack();
    }

    public class RubberDuck : IQuackable 
    {
        public void Quack() 
        {
            Console.WriteLine("Squeak");
        }
  
        public override string ToString() 
        {
            return "Rubber Duck";
        }
    }

    public class RedheadDuck : IQuackable 
    {
        public void Quack() 
        {
            Console.WriteLine("Quack");
        }
    }

    public class MallardDuck : IQuackable 
    {
         public void Quack() 
        {
            Console.WriteLine("Quack");
        }
 
        public override string ToString() 
        {
            return "Mallard Duck";
        }
    }

    public class DuckCall : IQuackable 
    {
         public void Quack() 
        {
            Console.WriteLine("Kwak");
        }
 
        public override string ToString() 
        {
            return "Duck Call";
        }
    }

    public class DecoyDuck : IQuackable 
    {
         public void Quack() 
        {
            Console.WriteLine("<< Silence >>");
        }
 
        public override string ToString() 
        {
            return "Decoy Duck";
        }
    }

    #endregion

    #region QuackCounter

    public class QuackCounter : IQuackable 
    {
        private IQuackable _duck;
  
        public QuackCounter(IQuackable duck) 
        {
            this._duck = duck;
        }
  
        public void Quack() 
        {
            _duck.Quack();
            Quacks++;
        }
 
        public static int Quacks{ get; private set; }

        public override string ToString() 
        {
            return _duck.ToString();
        }
    }

    #endregion

    #region Goose 

    public class GooseAdapter : IQuackable 
    {
        private Goose _goose;
 
        public GooseAdapter(Goose goose) 
        {
            this._goose = goose;
        }
  
        public void Quack() 
        {
            _goose.Honk();
        }
 
        public override string ToString() 
        {
            return "Goose pretending to be a Duck";
        }
    }

    public class Goose 
    {
        public void Honk() 
        {
            Console.WriteLine("Honk");
        }

        public override string ToString() 
        {
            return "Goose";
        }
    }
    #endregion
}
